test/index.test.js

```diff
@@ -0,0 +1,125 @@
+import sinon from 'sinon';
+import {
+  emptyDom,
+  testElm,
+  ButtonComponent,
+  DecoratedButton,
+  WrappedButton,
+} from './setup';
+import withStyles from '../src';
+
+describe('index.js:', () => {
+  const cache = {
+    get: sinon.stub(),
+    set: sinon.stub(),
+    delete: sinon.stub(),
+  };
+  const csjs = {
+    getCss: sinon.spy(),
+  };
+
+  before(() => {
+    const document = jsdom.jsdom(emptyDom);
+    global.document = document;
+    global.window = document.defaultView;
+  });
+
+  afterEach(() => {
+    cache.get.reset();
+    cache.set.reset();
+    cache.delete.reset();
+    csjs.getCss.reset();
+    withStyles.__ResetDependency__('cache');
+    withStyles.__ResetDependency__('csjs');
+  });
+
+  after(() => {
+    delete global.document;
+    delete global.window;
+  });
+
+  describe('withStyles', () => {
+    it('should return a valid React component that renders (decorator)', () => {
+      expect(ReactTestUtils.isElement(<DecoratedButton />)).to.be.true;
+
+      const rendered = ReactTestUtils.renderIntoDocument(<DecoratedButton />);
+      expect(ReactTestUtils.isCompositeComponent(rendered)).to.be.true;
+    });
+
+    it('should return a valid React component that renders (wrapped)', () => {
+      expect(ReactTestUtils.isElement(<ButtonComponent />)).to.be.true;
+      expect(ReactTestUtils.isElement(<WrappedButton />)).to.be.true;
+
+      const rendered = ReactTestUtils.renderIntoDocument(<WrappedButton />);
+      expect(ReactTestUtils.isCompositeComponent(rendered)).to.be.true;
+    });
+
+    it('should cache styles on subsequent mounts', () => {
+      withStyles.__Rewire__('cache', cache);
+      cache.get
+        .onFirstCall()
+        .returns(undefined)
+        .onSecondCall()
+        .returns({ style: testElm, count: 1 });
+
+      const rendered = ReactTestUtils.renderIntoDocument(<WrappedButton />);
+      rendered.componentWillMount();
+      expect(cache.set.calledOnce).to.be.true;
+    });
+
+    it('should clear styles cache on last component instance dismount', () => {
+      withStyles.__Rewire__('cache', cache);
+      cache.get
+        .onFirstCall()
+        .returns(undefined)
+        .onSecondCall()
+        .returns({ style: testElm, count: 1 });
+
+      const rendered = ReactTestUtils.renderIntoDocument(<WrappedButton />);
+      rendered.componentWillUnmount();
+      expect(cache.delete.calledOnce).to.be.true;
+    });
+
+    it('should not clear styles cache on normal component dismount', () => {
+      withStyles.__Rewire__('cache', cache);
+      cache.get
+        .onFirstCall()
+        .returns(undefined)
+        .onSecondCall()
+        .returns({ style: testElm, count: 1 })
+        .onThirdCall()
+        .returns({ styles: testElm, count: 2 });
+
+      const rendered = ReactTestUtils.renderIntoDocument(<WrappedButton />);
+      rendered.componentWillMount();
+      rendered.componentWillUnmount();
+      expect(cache.delete.called).to.false;
```

Comment:
Good catch, thanks.

It's being unmounted once, but it's been mounted twice (once in the initial `renderIntoDocument` call, and again in the `rendered.componentWillMount` call (manually triggered lifecycle method). So we expect the cache to remain in tact (until the last instance of it has been unmounted).